### PR TITLE
SQL-1117: Implement ProcessUri function

### PR DESCRIPTION
**Pull request by bucaojit**
Here's an example of what a unit test failure would look like ([link](https://evergreen.mongodb.com/task/mongo_powerbi_connector_windows_unit_test_patch_3a688e82b6085ef4502016e40e0823b5f9cf5f71_639911b2e3c331745f5d698f_22_12_13_23_58_57)): 
```
[2022/12/14 00:00:59.572] Failures Encountered:
[2022/12/14 00:00:59.572] "Result":"Failure ? ","Notes":"GetConnectionString - Simple URI","Details":" ([ DRIVER = \"aADF_ODBC_DRIVER\", DATABASE = \"db\", SERVER = \"localhost:27017\", AUTH_SRC = \"admin\" ] <> [ DRIVER = \"ADF_ODBC_DRIVER\", DATABASE = \"db\", SERVER = \"localhost:27017\", AUTH_SRC = \"admin\" ] )"
[2022/12/14 00:00:59.578] Command ''shell.exec' in "run unit-test"' failed: shell script encountered problem: exit code 1.
[2022/12/14 00:00:59.578] Task completed - FAILURE.
```

### CODE REVIEW

File: connector/MongoDBAtlasODBC.pq [hunk 1]
```diff
@@ -51,13 +47,73 @@ GetCredentialConnectionString = () as record =>
...
+        // http and https schemes automatically set port values to 80 and 443 respectively.
+        // The following checks whether the user has set the port values explicitly to 80 or 443.
+        // An unknown scheme will set the port value to -1 if it is not specified
+        Port =  if  (UriParts[Port] = 80 and not Text.Contains(uri, ":80")) or
+                    (UriParts[Port] = 443 and not Text.Contains(uri, ":443")) or
+                    UriParts[Port] = -1 then
+                    DefaultPort
+                else
+                    Number.ToText(UriParts[Port]),
+        SSL = NullFieldCheck(UriParts[Query][ssl]?, "SSL"),
```
Comment (bucaojit) [inline connector/MongoDBAtlasODBC.pq:89]: I was not able to figure out how to get these fields (`ssl`,`tls`,`authSource`) in a case insensitive way.  For example if the user had `authsource` all lowercase, it would not be used here.  I can create a ticket for that or do we think it's something we can address with documentation?

File: connector/MongoDBAtlasODBC.query.pq [hunk 2]
```diff
@@ -1,5 +1,233 @@
...
+
+GetExpectedError = (function as any) as text =>
+  let
+    Error = try function
+            catch (e) => e[Message]
+  in
+    Error;
+
+
+/// COMMON UNIT TESTING CODE
```
Comment (bucaojit) [inline connector/MongoDBAtlasODBC.query.pq:58]: This code from here to the end of the file is part of the [UnitTesting example](https://github.com/microsoft/DataConnectors/blob/master/samples/UnitTesting/UnitTesting.query.pq) and does not require review.

File: connector/MongoDBAtlasODBC.pq [hunk 3]
```diff
@@ -51,13 +52,71 @@ GetCredentialConnectionString = () as record =>
...
+                        DefaultAuthSrc,
+        Result = Record.AddField([], FieldName, AuthSrc)
     in 
         Result;
 
+ProcessUri = (uri as text) as any => 
+    let
+        DefaultPort = "27017",
+        UriParts = Uri.Parts(uri),
+        BadUri = UriParts[Host] = "" or UriParts[Fragment] <> ""
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.pq:81]: I am just wondering is extra parenthesis would help here.
Having 2 `=` following each others threw me off a bit. It looks like this is the expected syntax, so I'm thinking that if maybe it was written as `BadUri = (UriParts[Host] = "") or (UriParts[Fragment] <> "") ... it would help a little bit. Or maybe just having each condition on its own line. I don't know. Just not a big fan of this.

File: connector/MongoDBAtlasODBC.pq [hunk 4]
```diff
@@ -51,13 +52,71 @@ GetCredentialConnectionString = () as record =>
...
 
-// SQL-1117: Implement ProcessUri Function
-ProcessUri = (uri as text) as record => 
+NullFieldCheck = (field as any, fieldName as text ) =>
+    let 
+        Result = if field <> null then Record.AddField([], fieldName, field) else []
+    in
+        Result;
+
+ProcessAuthSrc = (queryEntry as any, pathEntry as any) =>
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.pq:61]: Should we actually bother doing any validation here? We could leave it for the driver to handle. WDYT?

File: connector/MongoDBAtlasODBC.pq [hunk 5]
```diff
@@ -51,13 +52,71 @@ GetCredentialConnectionString = () as record =>
...
+ProcessUri = (uri as text) as any => 
+    let
+        DefaultPort = "27017",
+        UriParts = Uri.Parts(uri),
+        BadUri = UriParts[Host] = "" or UriParts[Fragment] <> ""
+            or UriParts[UserName] <> "" or UriParts[Password] <> "",
+        // http and https schemes automatically set port values to 80 and 443 respectively.
+        // The following checks whether the user has set the port values explicitly to 80 or 443.
+        // An unknown scheme will set the port value to -1 if it is not specified
+        Port =  if  (UriParts[Port] = 80 and not Text.Contains(uri, ":80")) or
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.pq:86]: Is this a requirement for the connector to work? Does it complain about something like unknown scheme otherwise?

File: connector/MongoDBAtlasODBC.pq [hunk 6]
```diff
@@ -51,13 +52,71 @@ GetCredentialConnectionString = () as record =>
...
+                    (UriParts[Port] = 443 and not Text.Contains(uri, ":443")) or
+                    UriParts[Port] = -1 then
+                    DefaultPort
+                else
+                    Number.ToText(UriParts[Port]),
+        SSL = NullFieldCheck(UriParts[Query][ssl]?, "SSL"),
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.pq:92]: tls and ssl are the same option based on the doc. I see that you pass them in the array, I don't really know why we need to do that. Is there a field for SSL/TLS too? If not, can we let the driver take care of this? 
The [ssl](https://www.mongodb.com/docs/manual/reference/connection-string/#mongodb-urioption-urioption.ssl) option is equivalent to the [tls](https://www.mongodb.com/docs/manual/reference/connection-string/#mongodb-urioption-urioption.tls) option.

ODBC does not support Properties the way JDBC does.
SQLDriverConnect and SQLBrowseConnect use a connection string. SQLConnect needs a DSN.
If we can simply have something which matches the API, it would be much more easier to handle.
I understand that we might want add a field for the DB because it is more straightforward this way, so the gymnastic with authSource could still be required. Although, I'm not even sure since the DB passed here is used to setDatabase on the Rust client after establishing the connection.

**Review by nbagnard — COMMENTED** [review]
It looks good, but I have a few questions around what this means regarding the connection info requirement/usage.
I think if you time to discuss them offline with a live demo/code walk-through, that'd be nice and probably much faster :)

**Review by terakilobyte — COMMENTED** [review]
I second Natacha here. I wonder how much of the uri parsing we can push down into the driver to validate. If the spec changes in the future, we'd only need to upgrade the driver to validate any new options a user could pass in.

**Comment by bucaojit** [comment]
After talking with Natacha about these changes, we consulted with the team and decided to have the ODBC driver take in a URI option which will pass the mongoDB URI directly and have the Rust driver handle the parsing and option processing.  The ticket to address that is [SQL-1148](https://jira.mongodb.org/browse/SQL-1148).  The comments in this PR are related to the option processing of the connection string, I will update this PR once SQL-1148 is complete at which point most if not all of the processing will be unnecessary.

File: connector/MongoDBAtlasODBC.query.pq [hunk 7]
```diff
@@ -1,5 +1,201 @@
...
+
+GetExpectedError = (function as any) as text =>
+  let
+    Error = try function
+            catch (e) => e[Message]
+  in
+    Error;
+
+
+/// COMMON UNIT TESTING CODE - From DataConnectors/UnitTesting
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.query.pq:26]: I wish they had put that in there SDK 😞

File: connector/MongoDBAtlasODBC.query.pq [hunk 8]
```diff
@@ -1,5 +1,201 @@
...
+                                        funsig,
+                                        (item)=>
+                                            (if item[isOptional] then "optional " else "") &
+                                            Serialize.Identifier(item[Name]) & " as " & _serialize_typename(item[Value], true),
+                                        ", ") &
+                                     ")",
+
+        _serialize_function_return_type = (x) => _serialize_typename(Type.FunctionReturn(x), true), 
+
+        Serialize = (x) as text =>
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.query.pq:183]: I guess here is where we will have fun if we want to test our returned result set when we have MDB specific field.
I'm wondering which type it will end up being 🤔

File: connector/MongoDBAtlasODBC.query.pq [hunk 9]
```diff
@@ -1,5 +1,201 @@
...
-let
-    result = MongoDBAtlasODBC.Contents("", "")
-in
-    result
+section MongoDBAtlasODBCUnitTests;
+shared UnitTests = 
+[
+    facts = 
+    {
+        Fact("GetConnectionString - Sets URI option",
```
Comment (nbagnard) [inline connector/MongoDBAtlasODBC.query.pq:6]: Do you want to add a test where the DB is missing so that you can check that it fails with "Database value cannot be empty"?

**Review by nbagnard — CHANGES_REQUESTED** [review]
I have only one request to add a test checking that we get an error. If this does not make sense for any reason, let me know, that's Ok too.
Otherwise, it looks good :)

File: connector/MongoDBAtlasODBC.query.pq (hunk 7 above)
Comment (bucaojit) [reply connector/MongoDBAtlasODBC.query.pq:26]: Me too.

File: connector/MongoDBAtlasODBC.query.pq (hunk 9 above)
Comment (bucaojit) [reply connector/MongoDBAtlasODBC.query.pq]: Ah yes I had a test that did this previously, and removed it when I changed to using the URI option.  The empty database test is there now 👍

File: connector/MongoDBAtlasODBC.pq [hunk 10]
```diff
@@ -8,7 +8,7 @@ MongoDBAtlasODBCType = type function(
     mongodbUri as (type text meta[
         Documentation.FieldCaption = "MongoDB URI",
         Documentation.FieldDescription = "MongoDB connection URI",
-        Documentation.SampleValues = {"mongodb://hostname:port/admin"}
+        Documentation.SampleValues = {"mongodb://hostname:port/admin?ssl=true"}
```
Comment (terakilobyte) [inline connector/MongoDBAtlasODBC.pq:11]: do we also want to have `authSource=something` so that it's in line with the proposed changes for the jdbc driver?

Comment (bucaojit) [reply connector/MongoDBAtlasODBC.pq:11]: Would that be instead of the `admin` value, such as `mongodb://hostname:port/?ssl=true&authSource=admin` ?

Comment (terakilobyte) [reply connector/MongoDBAtlasODBC.pq:11]: I don't think so! I'm not sure, but I think we want the modal to look the same when it hits the UI. So the connection shown to users would be `mongodb://hostname:port?ssl=true&authSource=admin` and then the database is included from the info field, if I'm understanding it correctly.

Comment (bucaojit) [reply connector/MongoDBAtlasODBC.pq]: I updated the connection value to have the `authSource=admin` and removed the `admin` in the default db spot.  Yes there will be a `database` text field in the UI for the user to input the value.

**Review by nbagnard — APPROVED** [review]
LGTM :)
